### PR TITLE
Ht 39 start stop mapping

### PR DESCRIPTION
**Pull request by feenster**
- ability to lock a task for  mapping
- ability to stop mapping without chaning status with a comment
- ability to mark task as done with a comment

TODO
- Integration with editors
- Refactor when enhanced task api's are available

### CODE REVIEW

File: client/app/project/project.controller.js [hunk 1]
```diff
@@ -156,16 +202,100 @@
             var taskPromise = taskService.getTask(projectId, taskId);
             taskPromise.then(function (data) {
                 //task returned successfully
-                vm.selectedTask = data;
+                //vm.selectedTaskFeature = feature;
```
Comment (LindaAlblas) [inline client/app/project/project.controller.js]: remove

File: client/app/project/project.controller.js [hunk 2]
```diff
@@ -156,16 +202,100 @@
...
                 vm.isSelectTaskMappable = !data.taskLocked && (data.taskStatus === 'READY' || data.taskStatus === 'INVALIDATED');
+                vm.taskError = vm.isSelectTaskMappable ? '' : 'task-not-mappable';
                 vm.currentTab = 'mapping';
-                vm.mappingStep = 'view';
+                vm.mappingStep = 'viewing';
             }, function () {
                 // task not returned successfully
-                vm.selectedTask = null;
+                vm.selectedTaskData = null;
+                //vm.selectedTaskFeature = null;
```
Comment (LindaAlblas) [inline client/app/project/project.controller.js]: remove

File: client/app/project/project.controller.js [hunk 3]
```diff
@@ -156,16 +202,100 @@
...
                 vm.isSelectTaskMappable = false;
                 vm.currentTab = 'mapping';
-                vm.mappingStep = 'task-get-error';
+                vm.taskError = 'task-get-error';
+                vm.mappingStep = 'viewing';
+            });
+        }
+
+        function setLockedTask(task) {
```
Comment (LindaAlblas) [inline client/app/project/project.controller.js]: add comment - also, the function doesn't seem to get used? Or will it get used after the refactor?

File: client/app/project/project.controller.js [hunk 4]
```diff
@@ -156,16 +202,100 @@
...
+                }
+                else {
+                    //it's become unmappable while user was thinking
+                }
+            }, function () {
+                // TODO - may need to handle error
+            });
+        }
+
+        vm.unLockTask = function (comment, status) {
```
Comment (LindaAlblas) [inline client/app/project/project.controller.js]: missing comment

File: client/app/project/project.controller.js [hunk 5]
```diff
@@ -156,16 +202,100 @@
...
+            var taskId = vm.lockedTaskData.taskId;
+            var unLockPromise = taskService.unLockTask(projectId, taskId, comment, status);
+            unLockPromise.then(function (response) {
+                refreshProject(projectId);
+                onTaskSelection(taskService.getTaskFeatureById(vm.taskVectorLayer.getSource().getFeatures(), taskId));
+            }, function () {
+            });
+        }
+
+        vm.lockSelectedTask = function () {
```
Comment (LindaAlblas) [inline client/app/project/project.controller.js]: missing comment

File: client/app/services/style.service.js [hunk 6]
```diff
@@ -105,6 +105,7 @@
         function getSelectedStyleFunction(feature) {
             // get the base style for the feature and override it's stroke only.
             var baseStyle = getTaskStyleFunction(feature);
+            console.log(baseStyle.getFill().getColor());
```
Comment (LindaAlblas) [inline client/app/services/style.service.js]: remove

File: client/app/services/task.service.js [hunk 7]
```diff
@@ -43,6 +46,48 @@
             });
         }
 
+        function unLockTask(projectId, taskId, comment, status) {
```
Comment (LindaAlblas) [inline client/app/services/task.service.js]: add comment

File: client/app/services/task.service.js [hunk 8]
```diff
@@ -43,6 +46,48 @@
...
+                // when the response is available
+                return (response);
+            }, function errorCallback() {
+                // called asynchronously if an error occurs
+                // or server returns response with an error status.
+                return $q.reject("error");
+            });
+        }
+
+        function lockTask(projectId, taskId) {
```
Comment (LindaAlblas) [inline client/app/services/task.service.js]: add comment

File: client/app/services/task.service.js [hunk 9]
```diff
@@ -102,5 +147,26 @@
             }
             return candidates;
         }
+
+        function getTaskFeatureById(features, id) {
```
Comment (LindaAlblas) [inline client/app/services/task.service.js]: add comment

File: client/app/project/project.controller.js [hunk 10]
```diff
@@ -156,16 +202,100 @@
...
+            });
+        }
+
+        vm.unLockTask = function (comment, status) {
+            var projectId = vm.projectData.projectId;
+            var taskId = vm.lockedTaskData.taskId;
+            var unLockPromise = taskService.unLockTask(projectId, taskId, comment, status);
+            unLockPromise.then(function (response) {
+                refreshProject(projectId);
+                onTaskSelection(taskService.getTaskFeatureById(vm.taskVectorLayer.getSource().getFeatures(), taskId));
```
Comment (LindaAlblas) [inline client/app/project/project.controller.js]: this is a bit hard to read with a lot on the same line.

File: client/app/project/project.controller.js [hunk 11]
```diff
@@ -156,16 +202,100 @@
...
+            }, function () {
+            });
```
Comment (LindaAlblas) [inline client/app/project/project.controller.js]: is this going to stay empty? if so, remove

File: client/app/project/project.controller.js [hunk 12]
```diff
@@ -50,34 +57,51 @@
         }
 
         /**
+         * calculates padding number to makes sure there is plenty of clear space around feature on map to keep visual
+         * context of feature location
+         * @returns {number} - padding number
+         */
+        function getVPadding() {
```
Comment (feenster) [inline client/app/project/project.controller.js]: done

File: client/app/project/project.controller.js [hunk 13]
```diff
@@ -50,34 +57,51 @@
...
+            // padding to makes sure there is plenty of clear space around feature on map to keep visual
+            // context of feature location
+            return vm.map.getSize()[1] * 0.3;
+        }
+
+
+        function selectFeature(feature) {
```
Comment (feenster) [inline client/app/project/project.controller.js]: done

File: client/app/project/project.controller.js [hunk 14]
```diff
@@ -50,34 +57,51 @@
...
-                // padding to makes sure there is plenty of clear space around feature on map to keep visual
-                // context of feature location
-                var vPadding = vm.map.getSize()[1] * 0.3;
-                vm.map.getView().fit(feature.getGeometry().getExtent(), {padding: [vPadding, vPadding, vPadding, vPadding]});
+                selectFeature(feature);
             }
             else {
-                vm.selectedTask = null;
+                vm.selectedTaskData = null;
+                //vm.selectedTaskFeature = null;
```
Comment (feenster) [inline client/app/project/project.controller.js]: done

File: client/app/project/project.controller.js [hunk 15]
```diff
@@ -50,34 +57,51 @@
...
             }
         };
 
         /**
          * clears the currently selected task.  Clears down/resets the vm properties and clears the feature param in the select interaction object.
          */
         vm.clearCurrentSelection = function () {
-            vm.selectedTask = null;
+            vm.selectedTaskData = null;
+            //vm.selectedTaskFeature = null;
```
Comment (feenster) [inline client/app/project/project.controller.js]: done

File: client/app/project/project.controller.js [hunk 16]
```diff
@@ -91,26 +115,46 @@
...
                 addAoiToMap(vm.projectData.areaOfInterest);
-                addProjectTasksToMap(vm.projectData.tasks);
+                addProjectTasksToMap(vm.projectData.tasks, true);
             }, function () {
                 // project not returned successfully
                 // TODO - may want to handle error
             });
         };
 
+        function refreshProject(id) {
```
Comment (feenster) [inline client/app/project/project.controller.js]: done

File: client/app/project/project.controller.js [hunk 17]
```diff
@@ -91,26 +115,46 @@
...
+            }, function () {
+                // project not returned successfully
+                // TODO - may want to handle error
+            });
+
+        }
+
         /**
          * Adds project tasks to map as features from geojson
          * @param tasks
```
Comment (feenster) [inline client/app/project/project.controller.js]: done

File: client/app/project/project.controller.js (hunk 1 above)
Comment (feenster) [inline client/app/project/project.controller.js]: done

File: client/app/project/project.controller.js (hunk 2 above)
Comment (feenster) [inline client/app/project/project.controller.js]: done

File: client/app/project/project.controller.js (hunk 3 above)
Comment (feenster) [inline client/app/project/project.controller.js]: it's is called from html

Comment (feenster) [inline client/app/project/project.controller.js]: deleted, not used

File: client/app/project/project.controller.js (hunk 11 above)
Comment (feenster) [inline client/app/project/project.controller.js]: added to do placeholder

File: client/app/project/project.controller.js (hunk 5 above)
Comment (feenster) [inline client/app/project/project.controller.js]: done

File: client/app/services/task.service.js (hunk 7 above)
Comment (feenster) [inline client/app/services/task.service.js]: done

**Pull request merged**